bindings/azure/eventgrid/eventgrid.go

```diff
@@ -0,0 +1,186 @@
+					a.logger.Error(err.Error())
+				}
+			} else if string(ctx.Method()) == "POST" {
+				bodyBytes := ctx.PostBody()
+
+				a.logger.Debug(string(bodyBytes))
+				err = handler(&bindings.ReadResponse{
+					Data: bodyBytes,
+				})
+				if err != nil {
+					a.logger.Error(err.Error())
+				}
+			}
+		}
+	}
+
+	fasthttp.ListenAndServe(":8080", m)
+
+	a.logger.Debug("listening for Event Grid events at http://localhost:8080/api/events")
+
+	return nil
+}
+
+func (a *AzureEventGrid) Write(req *bindings.WriteRequest) error {
+	a.logger.Debug("Write() called. data: %s", string(req.Data))
+
+	request := fasthttp.AcquireRequest()
+	defer fasthttp.ReleaseRequest(request)
+	request.Header.SetMethod(fasthttp.MethodPost)
+	request.Header.Set("Content-Type", "application/cloudevents+json")
+	request.Header.Set("aeg-sas-key", a.metadata.ClientSecret)
+	request.SetRequestURI(a.metadata.TopicEndpoint)
+	request.SetBody(req.Data)
+
+	response := fasthttp.AcquireResponse()
+	defer fasthttp.ReleaseResponse(response)
+
+	client := &fasthttp.Client{WriteTimeout: time.Second * 10}
+	err := client.Do(request, response)
+	if err != nil {
+		a.logger.Error(err.Error())
+	}
+
+	if response.StatusCode() != fasthttp.StatusOK {
+		body := response.Body()
+		a.logger.Error(string(body))
+		return errors.New(string(body))
+	}
+
+	return nil
+}
+
+func (a *AzureEventGrid) parseMetadata(metadata bindings.Metadata) (*azureEventGridMetadata, error) {
+	b, err := json.Marshal(metadata.Properties)
+	if err != nil {
+		return nil, err
+	}
+
+	var eventGridMetadata azureEventGridMetadata
+	err = json.Unmarshal(b, &eventGridMetadata)
+	if err != nil {
+		return nil, err
+	}
+	return &eventGridMetadata, nil
+}
+
+func (a *AzureEventGrid) createSubscription() error {
+	clientCredentialsConfig := auth.NewClientCredentialsConfig(a.metadata.ClientID, a.metadata.ClientSecret, a.metadata.TenantID)
+
+	subscriptionClient := eventgrid.NewEventSubscriptionsClient(a.metadata.SubscriptionID)
+	authorizer, err := clientCredentialsConfig.Authorizer()
+	if err != nil {
+		return err
+	}
+	subscriptionClient.Authorizer = authorizer
+
+	scope := fmt.Sprintf("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.EventGrid/topics/%s", a.metadata.SubscriptionID, a.metadata.ResourceGroupName, a.metadata.TopicName)
+
+	eventInfo := eventgrid.EventSubscription{
+		EventSubscriptionProperties: &eventgrid.EventSubscriptionProperties{
+			Destination: eventgrid.WebHookEventSubscriptionDestination{
+				EndpointType: eventgrid.EndpointTypeWebHook,
+				WebHookEventSubscriptionDestinationProperties: &eventgrid.WebHookEventSubscriptionDestinationProperties{
+					EndpointURL: &a.metadata.SubscriberEndpoint,
+				},
+			},
+		},
+	}
+
+	a.logger.Debugf("Attempting to create or update Event Grid subscription. scope=%s endpointURL=%s", scope, a.metadata.SubscriberEndpoint)
+	result, err := subscriptionClient.CreateOrUpdate(context.Background(), scope, a.metadata.EventGridSubscriptionName, eventInfo)
```

Comment:
Should there not be WebHook ValidationCode handshake after that?
https://docs.microsoft.com/en-us/azure/event-grid/security-authentication#webhook-event-delivery

Reply:
No, that's the old way. We are using the CloudEvents method now which is what the OPTIONS callback is for... More information here: https://github.com/cloudevents/spec/blob/v1.0/http-webhook.md

Reply:
Thanks, learned something new - docs are outdated ....